subsys/net/lib/ftp_client/src/ftp_client.c

```diff
@@ -0,0 +1,793 @@
+		if (ret < 0) {
+			LOG_ERR("send(data) failed: %d", -errno);
+			ret = -errno;
+			break;
+		}
+		offset += ret;
+		ret = 0;
+	}
+
+	LOG_DBG("DATA sent");
+	close(data_sock);
+	return ret;
+}
+
+/**@brief Receive FTP message from socket
+ */
+int do_ftp_recv_ctrl(bool post_result, int success_code)
+{
+	int ret;
+	struct pollfd fds[1];
+
+	memset(ctrl_buf, 0x00, NET_IPV4_MTU);
+
+	/* Receive FTP control message */
+	fds[0].fd = client.sock;
+	fds[0].events = POLLIN;
+	ret = poll(fds, 1, MSEC_PER_SEC * CONFIG_FTP_LISTEN_TIME);
+	if (ret <= 0) {
+		LOG_ERR("poll(ctrl) failed: (%d)", -errno);
+		return -ETIMEDOUT;
+	}
+	ret = recv(client.sock, ctrl_buf, NET_IPV4_MTU, 0);
+	if (ret < 0) {
+		LOG_ERR("recv(ctrl) failed: (%d)", -errno);
+		return -errno;
+	}
+	if (ret == 0) {
+		/* Server close connection */
+		return -ECONNRESET;
+	}
+
+	LOG_HEXDUMP_DBG(ctrl_buf, ret, "RXC");
+
+	if (post_result) {
+		client.ctrl_callback(ctrl_buf, ret);
+	}
+
+	LOG_DBG("CTRL received");
+	return parse_return_code(ctrl_buf, success_code);
+}
+
+/**@brief Receive FTP data from socket
+ */
+int do_ftp_recv_data(const char *pasv_msg)
+{
+	int ret;
+	int data_sock;
+	struct pollfd fds[1];
+	char data_buf[NET_IPV4_MTU];
+
+	/* Establish data channel */
+	ret = establish_data_channel(pasv_msg);
+	if (ret <= 0) {
+		return ret;
+	}
+
+	/* Receive FTP data message */
+	data_sock = ret;
+	fds[0].fd = data_sock;
+	fds[0].events = POLLIN;
+	do {
+		memset(data_buf, 0x00, NET_IPV4_MTU);
+		ret = poll(fds, 1, MSEC_PER_SEC * CONFIG_FTP_LISTEN_TIME);
+		if (ret <= 0) {
+			LOG_ERR("poll(data) failed: (%d)", -errno);
+			break;
+		}
+		if ((fds[0].revents & POLLIN) != POLLIN) {
+			LOG_INF("No more data");
+			break;
+		}
+		ret = recv(data_sock, data_buf, NET_IPV4_MTU, 0);
+		if (ret < 0) {
+			LOG_ERR("recv(data) failed: (%d)", -errno);
+			break;
+		}
+		if (ret == 0) {
+			/* Server close connection */
+			break;
+		}
+
+		LOG_HEXDUMP_DBG(data_buf, ret, "RXD");
+		client.data_callback(data_buf, ret);
+	} while (true);
+
+	close(data_sock);
+	LOG_DBG("DATA received");
+	return 0;
+}
+
+void data_task(struct k_work *item)
+{
+	struct data_task_t *task_param =
+		CONTAINER_OF(item, struct data_task_t, work);
+
+	if (task_param->task == TASK_RECEIVE) {
+		do_ftp_recv_data(task_param->ctrl_msg);
+		k_sem_give(&rx_done);
+	} else if (task_param->task == TASK_SEND) {
+		do_ftp_send_data(task_param->ctrl_msg,
+				task_param->data,
+				task_param->length);
+		k_sem_give(&tx_done);
+	}
+}
+
+void keepalive_handler(struct k_work *work)
+{
+	int ret;
+
+	ret = do_ftp_send_ctrl(CMD_NOOP, sizeof(CMD_NOOP) - 1);
+	if (ret == 0) {
+		(void)do_ftp_recv_ctrl(false, FTP_CODE_200);
+	}
+}
+
+K_WORK_DEFINE(keepalive_work, keepalive_handler);
+
+void keepalive_timeout(struct k_timer *dummy)
+{
+	k_work_submit_to_queue(client.work_q, &keepalive_work);
+}
+
+K_TIMER_DEFINE(keepalive_timer, keepalive_timeout, NULL);
+
+/**@brief API to open FTP connection
```

Comment:
API functions are already documented in the header, no need to duplicate docs.

Reply:
All one-line comment removed